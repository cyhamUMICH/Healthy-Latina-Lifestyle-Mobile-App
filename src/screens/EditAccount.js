import React, { useEffect, useState } from 'react';
import { View } from 'react-native';
import { styles } from '../styles/Styles';
import { colors } from '../styles/Colors';
import { Button } from 'react-native-elements';
import { Image } from 'react-native';
import { Alert } from 'react-native';
import { TextInput } from 'react-native-gesture-handler';
import { Text } from 'react-native';
import firebase, { auth } from 'firebase';
import 'firebase/firestore';
import 'firebase/storage';


const EditAccount = (props) => {

const [name, setName] = useState('');
const [email, setEmail] = useState('');
const [username, setUsername] = useState('');

const [oldName, setOldName] = useState('');
const [oldEmail, setOldEmail] = useState('');
const [oldUsername, setOldUsername] = useState('');



useEffect(() =>{
    const loadOldName = async () =>{
    var user = firebase.auth().currentUser
    await firebase.firestore().collection('Users').doc(user.uid).get()
    .then(documentSnapshot => getName(documentSnapshot))
    .then(oldName => setOldName(oldName))

    }

    const loadOldUsername = async () =>{

        var user = firebase.auth().currentUser
            await firebase.firestore().collection('Users').doc(user.uid).get()
            .then(documentSnapshot => getUsername(documentSnapshot))
            .then(oldUsername => setOldUsername(oldUsername))

        }

    const loadOldEmail = async () =>{
               var user = firebase.auth().currentUser
                await firebase.firestore().collection('Users').doc(user.uid).get()
                .then(documentSnapshot => getEmail(documentSnapshot))
                .then(oldEmail => setOldEmail(oldEmail))
              }

    const getName = (documentSnapshot) =>{
    return documentSnapshot.get('name');

    }
     const getEmail= (documentSnapshot) =>{
        return documentSnapshot.get('email');
        }

const getUsername= (documentSnapshot) =>{
        return documentSnapshot.get('username');
        }

loadOldEmail();
loadOldName();
loadOldUsername();
})

function createTwoButtonAlertForDelete(){
Alert.alert(
      "Delete Account",
      "Are you sure you want to delete your account? This CANNOT be undone",
      [
        {
          text: "Cancel",
          style: "cancel"
        },
        { text: "OK", onPress: () => deleteAccount() }
      ],
      { cancelable: true }
    );
}

async function checkUsername(user)
{
var temp = true;
const load =async  (user)=>{//loads email from a given username


    await firebase.firestore().collection('Users').where('username', '==', user).get()
            .then(querySnapshot => {
                if (querySnapshot.size > 0) {
                temp = false;//this means that it is taken
                }
             })
    }
await load(user);
return temp;
}

function createTwoButtonAlertForUpdate(){
Alert.alert(
      "Update Account",
      "This will update all populated fields above. Are you sure?",
      [
        {
          text: "Cancel",
          style: "cancel"
        },
        { text: "OK", onPress: () => updateProfile()}
      ],
      { cancelable: true }
    );
}

function createOneButtonAlertForUnableToUpdateUsername(){
Alert.alert(
      "Username Already Taken",
      "Your desired username is already taken. Your Username was NOT updated",
      [
        { text: "OK"}
      ],
      { cancelable: true }
    );
}


function  checkValues(){
 if(name === "")
        {
        name => setName(oldName);
        }
        if(username)
        {
        username => setUsername(oldUsername);
        }
        if(email)
        {
        email => setEmail(oldEmail);
        }
        }



async function updateProfile(){

var user = firebase.auth().currentUser

console.log(user.uid)

var aName = name;
var aEmail = email;
var aUsername = username;
 if(aName === "")
 {
 aName = oldName;
 }
 if(aEmail === "")
 {
 aEmail = oldEmail;
 }
 if(aUsername === "")
 {
 aUsername = oldUsername;
 }
 else
 {
 const check = await checkUsername(aUsername);
    if(!check && aUsername != oldUsername)
    {
    aUsername =oldUsername;
    createOneButtonAlertForUnableToUpdateUsername();
    }
 }

//await checkValues();

const dbh = firebase.firestore().collection('Users').doc(user.uid)

await dbh.set({
name: aName,
username: aUsername,
email: aEmail
}, {merge:true});
user.updateEmail(aEmail);
user.updateProfile({
  displayName:aUsername
  });

props.navigation.navigate("Home", { navigation: props.navigation })

}


function deleteAccount(){//also deletes data under their user id from firestore
var user = firebase.auth().currentUser;


const dbh = firebase.firestore().collection('Users').doc(user.uid)
dbh.delete()

user.delete().then(function() {
  // User deleted.


}, function(error) {
  // An error happened.
});
props.navigation.navigate("Login")}



  return (
    <View style={styles.app}>
      <View style={styles.fullWidthWindow}>
      <View style={styles.smallerWidthWindow}>
        <Image 
        source={require('../../assets/temporary/Meditation3.png')} 
        style={styles.editAccountProfile}
        />

     
        <Button 
        buttonStyle={styles.button}
        titleStyle={styles.buttonText}
        title="Change profile picture"
        />


      <View style={styles.inputView}>
        <TextInput
          style={styles.inputText}
          textAlign="center"
          placeholder="name"
          placeholderTextColor={colors.text}
          onChangeText={name => setName(name)}
        />

        <TextInput
          style={styles.inputText}
          textAlign="center"
          placeholder="username"
          placeholderTextColor={colors.text}
          onChangeText={username => setUsername(username)}
        />
 
        <TextInput
          style={styles.inputText}
          textAlign="center"
          placeholder="email"
          placeholderTextColor={colors.text}
          onChangeText={email => setEmail(email)}
        />
      </View>

      <Button 
        buttonStyle={styles.button}
        titleStyle={styles.buttonText}
        title="Change password" onPress={() => props.navigation.navigate("PasswordReset")}
        />

      <Button 
        buttonStyle={styles.button}
        titleStyle={styles.buttonText}
        title="Update profile information" onPress={() => {createTwoButtonAlertForUpdate(); props.navigation.navigate("Home", { navigation: props.navigation })}}
        />
        <Button
                buttonStyle={styles.button}
                titleStyle={styles.buttonText}
                title="Delete Account" onPress={() => createTwoButtonAlertForDelete()}
                />
     </View>
     </View>
    </View>
  );
};

export default EditAccount;