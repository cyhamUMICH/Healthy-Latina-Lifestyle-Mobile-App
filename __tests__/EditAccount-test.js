// Code from: https://github.com/byCedric/expo-guide-ci/blob/master/__tests__/App-test.js
//      and https://jestjs.io/docs/tutorial-react-native
import React from 'react';
import EditAccount from '../src/screens/EditAccount';
import renderer from 'react-test-renderer';

test('Renders Edit Account With the Loading Screen', async () => {
  const tree = renderer.create(<EditAccount />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders Edit Account Without the Loading Screen', async () => {
  const tree = renderer.create(<EditAccount skipLoadingScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});