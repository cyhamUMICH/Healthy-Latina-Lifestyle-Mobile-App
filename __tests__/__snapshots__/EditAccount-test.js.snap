// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Edit Account With the Loading Screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": "2%",
        "paddingTop": "2%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "padding": "10%",
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "alignItems": "center",
            "aspectRatio": 1,
            "borderRadius": 100,
            "height": null,
            "justifyContent": "center",
            "marginBottom": "5%",
            "marginHorizontal": "15%",
            "width": null,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Change profile picture
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "marginVertical": "2%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          collapsable={false}
          onChangeText={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          placeholder="name"
          placeholderTextColor="#000"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#00A89D",
              "borderRadius": 30,
              "marginBottom": "2%",
              "paddingHorizontal": "5%",
              "paddingVertical": "2%",
            }
          }
          textAlign="center"
          underlineColorAndroid="transparent"
        />
        <TextInput
          allowFontScaling={true}
          collapsable={false}
          onChangeText={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          placeholder="username"
          placeholderTextColor="#000"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#00A89D",
              "borderRadius": 30,
              "marginBottom": "2%",
              "paddingHorizontal": "5%",
              "paddingVertical": "2%",
            }
          }
          textAlign="center"
          underlineColorAndroid="transparent"
        />
        <TextInput
          allowFontScaling={true}
          collapsable={false}
          onChangeText={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          placeholder="email"
          placeholderTextColor="#000"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#00A89D",
              "borderRadius": 30,
              "marginBottom": "2%",
              "paddingHorizontal": "5%",
              "paddingVertical": "2%",
            }
          }
          textAlign="center"
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Change password
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Update profile information
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Delete Account
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Renders Edit Account Without the Loading Screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": "2%",
        "paddingTop": "2%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "padding": "10%",
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "alignItems": "center",
            "aspectRatio": 1,
            "borderRadius": 100,
            "height": null,
            "justifyContent": "center",
            "marginBottom": "5%",
            "marginHorizontal": "15%",
            "width": null,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Change profile picture
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "marginVertical": "2%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          collapsable={false}
          onChangeText={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          placeholder="name"
          placeholderTextColor="#000"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#00A89D",
              "borderRadius": 30,
              "marginBottom": "2%",
              "paddingHorizontal": "5%",
              "paddingVertical": "2%",
            }
          }
          textAlign="center"
          underlineColorAndroid="transparent"
        />
        <TextInput
          allowFontScaling={true}
          collapsable={false}
          onChangeText={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          placeholder="username"
          placeholderTextColor="#000"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#00A89D",
              "borderRadius": 30,
              "marginBottom": "2%",
              "paddingHorizontal": "5%",
              "paddingVertical": "2%",
            }
          }
          textAlign="center"
          underlineColorAndroid="transparent"
        />
        <TextInput
          allowFontScaling={true}
          collapsable={false}
          onChangeText={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          placeholder="email"
          placeholderTextColor="#000"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#00A89D",
              "borderRadius": 30,
              "marginBottom": "2%",
              "paddingHorizontal": "5%",
              "paddingVertical": "2%",
            }
          }
          textAlign="center"
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Change password
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Update profile information
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#00A89D",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": "2%",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Delete Account
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
